### ISSUE TITLE
Remove Socat for local access to service containers

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#869]
Ideally we could have a fallback when `socat` is not installed. My idea is to inject a perl script in the container when `socat` isn't available that is capable of proxying connections. Why choose Perl? It is the only scripting language included in the default Ubuntu and Debian containers.

I found the an proxy Perl script here:
https://github.com/pkrumins/perl-tcp-proxy/blob/master/tcp-proxy.pl

I adjusted the below example to work with Redis:

```perl
#!/usr/bin/perl
use warnings;
use strict;

use IO::Socket;
use IO::Select;

my $ioset = IO::Select->new;
my %socket_map;

my $debug = 1;

sub new_conn {
    my ($host, $port) = @_;
    return IO::Socket::INET->new(
        PeerAddr => $host,
        PeerPort => $port
    ) || die "Unable to connect to $host:$port: $!";
}

sub new_server {
    my ($host, $port) = @_;
    my $server = IO::Socket::INET->new(
        LocalAddr => $host,
        LocalPort => $port,
        ReuseAddr => 1,
        Listen    => 100
    ) || die "Unable to listen on $host:$port: $!";
}

sub new_connection {
    my $server = shift;
    my $client = $server->accept;

    my $remote = new_conn($ENV{"REDIS_PORT_6379_TCP_ADDR"}, $ENV{REDIS_PORT_6379_TCP_PORT});
    $ioset->add($client);
    $ioset->add($remote);

    $socket_map{$client} = $remote;
    $socket_map{$remote} = $client;
}

sub close_connection {
    my $client = shift;
    my $remote = $socket_map{$client};
    
    $ioset->remove($client);
    $ioset->remove($remote);

    delete $socket_map{$client};
    delete $socket_map{$remote};

    $client->close;
    $remote->close;
}

my $server = new_server('0.0.0.0', $ENV{REDIS_PORT_6379_TCP_PORT});
$ioset->add($server);

while (1) {
    for my $socket ($ioset->can_read) {
        if ($socket == $server) {
            new_connection($server);
        }
        else {
            next unless exists $socket_map{$socket};
            my $remote = $socket_map{$socket};
            my $buffer;
            my $read = $socket->sysread($buffer, 4096);
            if ($read) {
                $remote->syswrite($buffer);
            }
            else {
                close_connection($socket);
            }
        }
    }
}
```

So how would this work? We would embed the perl script in our `build.sh` file and write that file from inside the container. It could look something like this:

```
echo IyEvdXNyL2Jpbi9wZXJsCnByaW50ICJoZWxsbyB3b3JsZCIKCg== | base64 -d > /tmp/drone-proxy
```

We could then invoke with `/tmp/drone-proxy $IP $PORT` as an alternative to `socat`.

It may seem a bit strange to embed the file in the `build.sh` script, but this is because when #749 is implemented we won't be able to inject files into the container any other way.

**Comment by bradrydzewski** [comment]
updated script. It can act as a polyfill (so to speak) for socat

```perl
#!/usr/bin/perl
use IO::Socket;
use IO::Select;

$input = join(' ',@ARGV);
$input =~ m/TCP-LISTEN:(.*),fork TCP:(.*):(.*)/;
$addr = $2;
$port = $3;

my $ioset = IO::Select->new;
my %socket_map;

my $server = IO::Socket::INET->new(
	LocalAddr => '0.0.0.0',
	LocalPort => $port,
	ReuseAddr => 1,
	Listen	=> 100
);
$ioset->add($server);

while (1) {
	for my $socket ($ioset->can_read) {
		if ($socket == $server) {
			my $client = $server->accept;
			my $remote = IO::Socket::INET->new(
				PeerAddr => $addr,
				PeerPort => $port
			);
			$ioset->add($client);
			$ioset->add($remote);
			$socket_map{$client} = $remote;
			$socket_map{$remote} = $client;
		}
		else {
			next unless exists $socket_map{$socket};
			my $remote = $socket_map{$socket};
			my $buffer;
			my $read = $socket->sysread($buffer, 4096);
			$remote->syswrite($buffer);
		}
	}
}
```

```sh
echo H4sICGKv1VQAA3NvY2F0LnBsAH1SXUvDQBB8Tn7FipUmkpr6gWBKgyIiBdGixVeJ6RZP00u4S6wi8be7t3exFsWEhNzO7M7MXba34kar+FHIuEJV+I1GmNwkyV2Zv2A9Wq+xwJzWfk/IqqlhDM+lkEEf+tHp2e3lfTj6Rj5hGc/Op4Oryd3s4joJ9nbDaFGqF6Air/gVU0M2nyua1Dug76pUZmrvkDSW79ATpUZTWIsPUomrkQF3NLt7WGaVY2tUr6g6OqNJMrm+mHFT4HtXZZ4VZ6yXQn+4x3c/csCUxVNgF1S8RcrdsfcNS+gapWdWw6HPYY2/QUoRAqdOVX/1JAqEYD+ED9+j0MDm2A8EXU+eyQeF2ZxJnlgQ4ijjcRfFYp5pzwuBkvfGQiSa51jRYTiCwmVZ4z/h6Zoiqi4Q73v0Xd4Ib6ohT95IaD38AVhtB6yP5cN1tMa25fym2DpTLNtQWnqwoL+O80t8q6GRBWoN+EaHoGFjhP1uf2/Fv6zHZrFA9aMpm69bBql+16YUOF4ER8OTYxfRCjBnpUSNHSl03lu/9b8ACaSZylQDAAA= | base64 -d | gunzip > /tmp/socat && chmod +x /tmp/socat
```

**Comment by ulrichSchreiner** [comment]
hi,

as long as all of your containers run on the same docker-host you could start the containers with `--net=container:<containerid>` and all of the containers share the same IP-stack. so all containers can access the others via "localhost" :-)

only an idea ...

**Comment by bradrydzewski** [comment]
wow, I had no clue that setting even exists. Thanks for the tip! I'll definitely scrap this change and implement what you've suggested

**Comment by shawnzhu** [comment]
@ulrichSchreiner given a scenario where more than one service container, like a build requires both MySQL and CouchDB, then it requires drone starts one service container first (ex. MySQL), then start a CouchDB container and the build container by sharing MySQL container's ip address and port numbers.

So it requires careful configuration of all service images to avoid port conflicts because all processes in containers are going to share same ip address and port numbers.

**Comment by ulrichSchreiner** [comment]
@shawnzhu you're right. There could be port conflicts. But afaik the current solution uses `socat` to tunnel the port of another image to the same port of the build container. So i think this is not a new problem.

You could start the service containters before the build container and use the docker flag `--ad-host` to put entries in `/etc/hosts` so your services are available as hosts and you won't have port conflicts. But then you need fixed names for the serivce hosts and these names could collide. And you also have to have the correct order of starting and injecting the hostnames.

You can also have some *ambassador* container which is always started at first and which does the complete wiring, but ...

Imho drone should be a CI server. If you want to fire up a whole network of docker containers you should use `docker compose`. Or use [weave](https://github.com/zettio/weave) or [socketplane](http://socketplane.io/) or [flannel](https://github.com/coreos/flannel). But i don't think drone is the right tool to configure some sort of SDN.

Or you should be careful with the selection of your containers :-)

**Comment by shawnzhu** [comment]
@ulrichSchreiner I would say tunnel the port of another container to the build container (iptables NAT rule) is different from sharing an ip namespace among different containers from networking perspective.

yes, we should be careful with the selection/make of service images to avoid possible port/hostnames conflict, but docker itself should guarantee the hostnames conflict like when linking a build container to two couchdb service containers from same image (although this is not a drone feature yet).

**Comment by dave-tucker** [comment]
Re: use of `--net=container:foo` I don't think this should be the default behaviour due to risk of port conflicts... This is likely if you require two or more versions of the same service container for your application integration tests (this point is moot when matrix builds are supported). 

Sharing a namespace should however be an option as I suggested in #865 but I see the main usage for this is to cover corner cases. For example, my `openvswitch` service creates ports in the container netns, which the container running the tests should be able to see using Netlink.

[Docker Links](http://docs.docker.com/userguide/dockerlinks/) is the best default IMO. It's pretty easy to write tests that expect the database to be accessible from `//postgres:5432` or `//mysql:3306` which is handled by adding the alias to `/etc/hosts/`. Alternatively one can check for the presence of environment variables like `MYSQL_PORT_3306_TCP`

I don't think the tunnelling ports to appear on `localhost` adds anything as the above seem to work fine for me. 

Removing the `socat` dependency means that the Docker Language Stacks should work as expected.
Replacing it with "magic" in the CI system (e.g injecting a script) means it's harder to repeat on developer machines.

Thats the killer feature in Drone for me. I can have `make check` run inside docker containers on the developers machine, and have the exact same thing run in Drone.

**Comment by bradrydzewski** [comment]
closing in favor of better fix in 0.4
